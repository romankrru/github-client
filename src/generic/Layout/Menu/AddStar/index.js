// @flow
import React, {useState, Fragment, useEffect} from 'react';
import produce from 'immer';
import {Button, Confirm} from 'semantic-ui-react';
import {type ApolloQueryResult} from 'react-apollo';
import {useQuery, useMutation} from 'react-apollo-hooks';
import {loader} from 'graphql.macro';

const query = loader('./gql/query.graphql');
const addStarMutation = loader('./gql/add-star.graphql');
const removeStarMutation = loader('./gql/remove-star.graphql');

type TRepository = {
	id: string,

	stargazers: {
		nodes: Array<{login: string, id: string}>,
	},
};

type TViewer = {
	id: string,
	login: string,
};

type TCache = {
	viewer: TViewer,
	repository: TRepository,
};

const AddStar = (props: $Exact<{}>) => {
	const [isConfirmationShown, setIsConfirmationShown] = useState(false);
	const [isUserInformed, setIsUserInformed] = useState(false);
	const [isStarred, setIsStared] = useState(false);

	const queryResult: ApolloQueryResult<{
		repository?: TRepository,
		viewer?: TViewer,
	}> = useQuery(query);

	const repo = queryResult.data.repository;
	const viewer = queryResult.data.viewer;

	const addStar = useMutation(addStarMutation, {
		variables: {
			id: repo && repo.id,
		},

		optimisticResponse: {
			addStar: {
				starrable: {
					id: repo && repo.id,
					__typename: 'Repository',
				},

				__typename: 'AddStarPayload',
			},
		},

		update: (proxy, mutationResult) => {
			const dataFromCache: TCache = proxy.readQuery({query: query});

			const updated = produce(dataFromCache, draft => {
				draft.repository.stargazers.nodes.push({
					id: dataFromCache.viewer.id,
					login: dataFromCache.viewer.login,
					__typename: 'User',
				});
			});

			proxy.writeQuery({query: query, data: updated});
		},
	});

	const removeStar = useMutation(removeStarMutation, {
		variables: {
			id: repo && repo.id,
		},

		optimisticResponse: {
			removeStar: {
				starrable: {
					id: repo && repo.id,
					__typename: 'Repository',
				},

				__typename: 'RemoveStarPayload',
			},
		},

		update: (proxy, mutationResult) => {
			const dataFromCache: TCache = proxy.readQuery({query: query});

			const updated = produce(dataFromCache, draft => {
				draft.repository.stargazers.nodes.splice(
					draft.repository.stargazers.nodes.findIndex(
						stargazer => stargazer.id === dataFromCache.viewer.id,
					),
					1,
				);
			});

			proxy.writeQuery({query: query, data: updated});
		},
	});

	const onClick = () => {
		if (isStarred) {
			removeStar();
			setIsUserInformed(true);
			return;
		}

		if (!isUserInformed) {
			setIsConfirmationShown(true);
			return;
		}

		addStar();
	};

	const onConfirm = () => {
		setIsConfirmationShown(false);
		addStar();
	};

	useEffect(() => {
		if (!isUserInformed && isConfirmationShown) setIsUserInformed(true);
	}, [isConfirmationShown, isUserInformed]);

	useEffect(() => {
		if (!repo || !viewer) return;

		const isStarred = repo.stargazers.nodes.some(
			stargazer => stargazer.id === viewer.id,
		);

		setIsStared(isStarred);
	}, [repo]);

	return (
		<Fragment>
			<Button
				onClick={onClick}
				circular
				icon="star"
				color={isStarred ? 'yellow' : 'grey'}
			/>

			<Confirm
				header="Add star"
				content="This action will add star for this project on Github. Are you sure?"
				open={isConfirmationShown}
				onCancel={() => setIsConfirmationShown(false)}
				onConfirm={onConfirm}
			/>
		</Fragment>
	);
};

export default AddStar;
