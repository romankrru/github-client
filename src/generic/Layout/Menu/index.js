import React from 'react'
import { Menu } from 'semantic-ui-react'
import { NavLink, Link } from 'react-router-dom';


const MainMenu = props => {
    console.log(props)

    return (
        <Menu>
            <Menu.Item as={NavLink} to="/" name="Home" />
            <Menu.Item as={NavLink} to="/discover" name="Discover"/>

            <Menu.Menu position="right">
                <Menu.Item as={Link} to="/logout" name="Logout" />
            </Menu.Menu>
        </Menu>
    );
};

export default MainMenu;