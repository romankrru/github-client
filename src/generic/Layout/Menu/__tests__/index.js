import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Menu from '../';

describe('Menu', () => {
    it('should render correctly', () => {
        const output = shallow(<Menu />);

        expect(shallowToJson(output)).toMatchSnapshot();
    });
});
