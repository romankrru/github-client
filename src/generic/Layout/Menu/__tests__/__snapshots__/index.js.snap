// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render correctly 1`] = `
<Menu>
  <MenuItem
    as={[Function]}
    exact={true}
    name="Home"
    to="/"
  />
  <MenuItem
    as={[Function]}
    name="Discover"
    to="/discover"
  />
  <MenuMenu
    position="right"
  >
    <MenuItem
      as={[Function]}
      name="Logout"
      to="/logout"
    />
  </MenuMenu>
</Menu>
`;
